File: agents/python/src/lib/state.py
"""
This is the state definition for the AI.
It defines the state of the agent and the state of the conversation.
"""

from typing import List, Literal, NotRequired, TypedDict
from langgraph.graph import MessagesState

class Resource(TypedDict):
    """
    Represents a resource. Give it a good title and a short description.
    Can be either a web resource or a Tako chart.
    """
    url: str
    title: str
    description: str
    resource_type: NotRequired[Literal["web", "tako_chart"]]
    card_id: NotRequired[str]  # Tako card identifier for visualizations
    iframe_html: NotRequired[str]
    source: NotRequired[str]
    embed_url: NotRequired[str]  # Tako embed URL for iframe src

class DataQuestion(TypedDict):
    """
    Represents a structured data question with search configuration.
    """
    question: str
    search_effort: Literal["fast", "deep"]
    query_type: Literal["basic", "complex", "prediction_market"]

class Log(TypedDict):
    """
    Represents a log of an action performed by the agent.
    """
    message: str
    done: bool

class AgentState(MessagesState):
    """
    This is the state of the agent.
    It is a subclass of the MessagesState class from langgraph.
    """
    model: str
    research_question: str
    report: str
    resources: List[Resource]
    logs: List[Log]
    data_questions: NotRequired[List[DataQuestion]]
    explore_context: NotRequired[str]

File: agents/typescript/src/state.ts
import { Annotation } from "@langchain/langgraph";
import { CopilotKitStateAnnotation } from "@copilotkit/sdk-js/langgraph";

// Define a Resource annotation with properties for URL, title, and description
// Extended with Tako chart support
const ResourceAnnotation = Annotation.Root({
  url: Annotation<string>,
  title: Annotation<string>,
  description: Annotation<string>,
  content: Annotation<string>,
  resource_type: Annotation<'web' | 'tako_chart'>,
  card_id: Annotation<string | undefined>,
  iframe_html: Annotation<string | undefined>,
  source: Annotation<string>,
});

// Define a Log annotation with properties for message and done status
const LogAnnotation = Annotation.Root({
  message: Annotation<string>,
  done: Annotation<boolean>,
});

// Define the AgentState annotation, extending MessagesState
export const AgentStateAnnotation = Annotation.Root({
  model: Annotation<string>,
  research_question: Annotation<string>,
  report: Annotation<string>,
  resources: Annotation<(typeof ResourceAnnotation.State)[]>,
  logs: Annotation<(typeof LogAnnotation.State)[]>,
  data_questions: Annotation<string[]>,
  ...CopilotKitStateAnnotation.spec,
});

export type AgentState = typeof AgentStateAnnotation.State;
export type Resource = typeof ResourceAnnotation.State;
